lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -144,54 +192,71 @@ def _on_index_requested(self, event: IndexRequestedEvent) -> None:
-                groups.
+            index: the index to which the users must be granted access
+            extra_user_roles: the level of permissions that the user should be given.
 
         Raises:
             OpenSearchUserMgmtError if user creation fails
         """
-        roles = access_control.get("roles")
-        permissions = access_control.get("permissions")
-        action_groups = access_control.get("action_groups")
-
-        if permissions or action_groups:
-            # combine action groups and perms into a new role for this relation.
-            try:
-                self.user_manager.create_role(
-                    role_name=username,
-                    permissions=permissions,
-                    action_groups=action_groups,
-                )
-                roles.append(username)
-            except OpenSearchUserMgmtError as err:
-                logger.error(err)
-                raise
-
         try:
+            # Create a new role for this relation, encapsulating the permissions we care about. We
+            # can't create a "default" and an "admin" role once because the permissions need to be
+            # set to this relation's specific index.
+            self.user_manager.create_role(
+                role_name=username,
+                permissions=self.get_extra_user_role_permissions(extra_user_roles, index),
+            )
+            roles = [username]
             self.user_manager.create_user(username, roles, hashed_pwd)
-            if roles:
-                self.user_manager.patch_user(
-                    username,
-                    [{"op": "replace", "path": "/opendistro_security_roles", "value": roles}],
-                )
+            self.user_manager.patch_user(
+                username,
+                [{"op": "replace", "path": "/opendistro_security_roles", "value": roles}],
+            )
         except OpenSearchUserMgmtError as err:
             logger.error(err)
             raise
 
+    def get_extra_user_role_permissions(self, extra_user_roles: str, index: str) -> Dict[str, any]:
+        """Get relation role permissions from the extra_user_roles field.
+
+        Args:
+            extra_user_roles: role requested by the requirer unit, provided in relation databag.
+                This needs to be one of "admin" or "default", or it will be set to "default".
+                TODO should this fail and raise an error instead so provider charm authors can
+                guarantee they're getting the perms they expect?
+            index: if these permissions are index-specific, they will be assigned to this index.
+
+        Returns:
+            A dict containing the required permissions for the requested role.
+        """
+        if extra_user_roles.upper() not in ExtraUserRolePermissions._member_names_:
+            extra_user_roles = "default"
+
+        permissions = ExtraUserRolePermissions[extra_user_roles.upper()].value
```

Comment:
extra_user_roles should be a comma-separated list.
It is ok to solve in separate PR

Reply:
I've added this in, let me know what you think